### PR TITLE
[RFC22, RFC78] NixOS a la carte (proof of concept)

### PR DESCRIPTION
**Pull request by roberth**
###### Motivation for this change

 - RFC22: a minimal NixOS module list to improve performance. A 3× eval performance improvement is realistic [based on](https://github.com/NixOS/nixpkgs/issues/137168#issue-991981735) earlier experimentation on the postgresql test.
 - RFC78: independent modules that can be composed (almost) like functions, for use outside of conventional NixOS.
 - Allow alternative implementations to coexist

These are big goals that this PR does not achieve by itself, but works towards. The actual goal for this PR is to provide an example of the kinds of refactoring that can be done to make NixOS more modular.

What it does already achieve, is an example of how to build a `dockerTools` image using the NixOS `/etc` code:

```nix
    let
      nixosCore = (lib.nixos.core ({ config, modules, ... }: {
        imports = [ pkgs.nixosModule modules.etc ];
        environment.etc."hosts" = {
          text = ''
            127.0.0.1 localhost
            ::1 localhost
          '';
          mode = "0444";
        };
      }));
    in pkgs.dockerTools.streamLayeredImage {
      name = "poc";
      tag = "latest";
      fakeRootCommands = ''
        mkdir -p /etc
        ${nixosCore.config.system.build.etcActivationCommands}
      '';
      config.Cmd = pkgs.writeScript "poc-cmd" ''
        #!${pkgs.busybox}/bin/sh
        ${pkgs.busybox}/bin/cat /etc/hosts
      '';
    }
```

So there you have it, reusing a bit of NixOS with great evaluation performance. When more modules are refactored this way, it becomes actually useful, solving a real problem in practical `dockerTools` use such as https://github.com/NixOS/nixpkgs/issues/94636) and replacing incomplete solutions like #105685.

I've found that the best workflow for this refactoring is test driven development.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
@ofborg test docker-tools etc nixpkgs openssh

**Comment by pennae** [comment]
interesting! we've been trying to get to a faster nixos another way, by marking most modules as "optional" and having the module system not evaluate them after the initial import of their nix file unless a configurable path in `config` has values (in our plan this path would usually be a module enable option). the docs work we've been doing is part of this, since docs currently evaluate all modules anyway and users probably will not be very happy if most of the configuration.nix manpage disappears because they've selected only a subset of modules for their system :<

**Comment by roberth** [comment]
@pennae That's also a good approach; a bit more pragmatic, but seems to need a more complicated module system and does not necessarily stimulate improvements towards RFC78 for example. Could you show what changes you had to make?

**Comment by pennae** [comment]
@roberth we don't have anything for this yet since we're concentrating on doing the docs bit first (which in itself yields about a 40% improvement in system build time on this machine). this being a highly experimental endeavour we just haven't yet had the energy to spend on doing things that might end up rejected.

the idea was to tag each module that isn't essential with an extra toplevel key that describes properties of the modules, leading to eg
```
{ lib, config, ... }:

{
  module.enabledBy = [ "services" "theThing" "enable" ];

  options.services.thisThing.enable = mkEnableOption "the thing";

  # ...
}
```

the module system imports all modules as before, but does not such optional modules except for their `enabledBy`. it then checks the resulting `options` (not `config`!) for enabler paths and adds those to the evaluation. repeat until fixed point is reached. for modules that enable other modules we'd go ahead and allow dependency declarations, eg `module.requires` to avoid unnecessary iterations. our testing has shown that module parse time is about 5% of the total instantiation time for our configuration, and if nixos modules declare their dependencies correctly we'd only have to do a single iteration to enable all necessary modules.

**Comment by roberth** [comment]
> ~40% build time reduction.

You should be able to get about 3× faster, which is your reduction applied _twice_. That discrepancy can be explained by the choice of test and I was quite radical with my pruning, even going into some modules to set values to sensible defaults in order to remove more modules. It doubt whether those wins are feasible without refactoring those modules.

**Comment by pennae** [comment]
oh yes, a lot of the remaining time was most likely due to a large `environment.systemPackages` from our config. we had also pruned the config itself for a different test and got a 4x improvement with the pruned config. there's clearly a lot of improvement still possible!

**Comment by aanderse** [comment]
@roberth awesome! I'm looking forward to seeing more examples so I can better wrap my mind around this.
@pennae also awesome! I hope to hear more about that coming up.

File: nixos/tests/postgresql-container.nix [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    testScript = ''
+      def check_count(statement, lines):
+          return 'test $(psql --host=127.0.0.1 --username=postgres postgres -tAc "{}"|wc -l) -eq {}'.format(
+              statement, lines
+          )
+
+
+      machine.start()
+      machine.wait_for_unit("docker.socket")
+      machine.succeed("${pkgs.dockerTools.examples.postgresql} | docker load")
```
Comment (lheckemann) [inline nixos/tests/postgresql-container.nix:52]:
```suggestion
      machine.succeed("cat ${pkgs.dockerTools.examples.postgresql} | docker load")
```

**Comment by roberth** [comment]
I'm closing this because the work now happens in other PRs, which tend to be linked here.

**Comment by roberth** [comment]
**Tracking which changes have been cleaned up and merged**

Note to self: might want to bundle up some of the similar and simple changes, like module extractions and adding `imports`.

- #153211
  - [x] 18672bf2751bbae1d7eeec59f5f82229b79b8873 lib.nixos: init
  - [x] 62e7f0eda1c5acf0beb13a00a23f577912a6b8eb nixos/nixpkgs.nix: Make independent
- https://github.com/NixOS/nixpkgs/pull/155892 (`pkgsModule`)
   - [x] dd6d8e3f0c38370ac8fb082ea7a3be06c949deed pkgs.nixosModule: init
   - [x] 56c283e5c8dfd4d8d5daa15dfa3896beef2ac012 nixos/etc.nix: Make independent
- [x] 7f129e32362b0eeb439c811ba286c2bbc79b151c nixos/build.nix: Make independent
- [ ] 9b2af8673be82d48ce76c8c152de85ad921d26ba dockerTools: Add example of using NixOS' etc
- [ ] 4ec415cff9bcbaef08e0d900406a5c7181e71881 nixos/top-level.nix: Make extensible
- [ ] baa96b8b592092af7700fe40700d3e12eb4f3a74 nixos/system-with-activation.nix: Extract module
- [ ] 8da928c46b6b8f0035d03939b6ff0eaa3b31ce08 nixos/bootable.nix: Extract module
- [ ] 2bf147c7c798e7df033c0cd7af18e46eb9155f8b nixos/specialisation.nix: Extract module
- [ ] c78d41a3ead8fda3f7230d63f0d72fba77891a3f nixos: Move top-level stage2 def to stage-2.nix
- [ ] c8da2d49cffd9ba004f095afe12a8b8df30cc195 nixos: Move system.name logic to network-interfaces.nix
- [ ] c6ea188fb9ebaa5622f8976616549787b03c246c nixos/system-path-core.nix: Extract module
- [ ] dc2b3962c3de7d8f264d5a8b583f74c7dce3359a nixos/users-groups.nix: Make independent
- [ ] 42b59e3fa23d3b4c9831e4da8de3ddd462034b78 nixos/update-users-groups.pl: Fall back to root group for shadow
- [ ] b9cb1828875635c04c264653e2275b5bef2454f2 nixos: Move specialfs activation script to filesystems.nix
- [ ] db05855521be357be0058e8318e7fba3d720b194 nixos/top-level.nix: Move configurationName to grub.nix
- [ ] 8313fd36f4e71da13ff516ba109862f7afa3fa83 nixos/top-level.nix: Make independent of systemd.package
- [ ] 87724859d07b45c35bb8e56dadc67c69eb9dbd3b nixos/user-profiles.nix: Extract module
- [ ] f2cd71449b87035f98ea7883e2e07b0a879627bf nixos/systemd-user-activation.nix: Extract module
- [ ] a7a1438c1539aef4f4b7ba4046be010359415565 nixos/top-level.nix: Make independent of boot modules
- [ ] 7c0c2e2842d1b7929ddca6bd8a3b015f58475da7 nixos: Add system.activation.externalActivationScript
- [ ] 413be2c77e68fde0476a9154387b2f14e18a2ba2 nixos/top-level.nix: Declare dependencies
- [ ] b692a9239abb5ff197835388b36dc2245530d864 nixos/users-groups.nix: Declare dependencies
- [ ] cc748856af6ce75af0ab802451f9980e4d327eca nixos/activation-script: Declare dependencies
- [ ] 226d20e3e775047aeb9dc2b91a298f9f2aedcd82 lib.nixos: Add users module
- [ ] 77e8c73303aa3a087f957893db3714540f7dd129 dockerTools.streamLayeredImage: Make robust against cd in fakeRootCommands
- [ ] bc6808e8987efceea6a0d0933a5083f6b351281d nixos/nixos-core.nix: Add etcActivation module
- [ ] eb3fc9a07da1eec35c6974a66eded9623136e6a7 dockerTools: Update etc example
- [ ] 6adab4ce541a282d5a2a5bcc21c0e625d7b5ff20 nixos/top-level.nix: Add system.checks
- [ ] 25093d76c8a283f56779cdbf82c51475f2dc2b40 nixos/binsh.nix: Extract module
- [ ] 2f3e9f4dde8d95254e9e2cfe08857dcde8356384 nixos/environment-variables.nix: Extract module
- [ ] d9e94baeec65057cb67d8ee023490a042dd8ce82 (hercules-ci/nixos-a-la-carte-poc-2) nixos/postgresql.nix: Split module, create container, test it

------

related changes
  - #156533
  - #156503

**Comment by roberth** [comment]
Latest PR is for having a separate activation script, that can run _during_ the build of an image
 - https://github.com/NixOS/nixpkgs/pull/237040

**Comment by roberth** [comment]
More progress is blocked on having better CI and unit tests to make sure that modular usage and basic properties don't regress.

This involves making CI (ofborg or GHA) find the right test cases to run for the module(s) that are changed in a given PR.
It'd have to be more file based because evaluating all tests is too resource intensive for PR testing.

**Pull request closed without merging**